### PR TITLE
Fetch reddit posts for a users homepage from the reddit api closes #27

### LINKED ISSUE #27: Fetch reddit posts for a users homepage from the reddit api
**User Story**
As a backend developer, I'd like to fetch users home page/feed from the Reddit api, so that this data can then be mapped to a common data type and combined with their other feeds

**Acceptance Criteria**
- The posts retrieved must be specific to the user and the subreddits they are subscribed to.
- The user credentials should be stored in a config file, for now, which should NOT be committed.
- This config store must be generic such that it can be reused by other API clients 
- Changes must be unit-tested

**Notes**

### PR DESCRIPTION
**Pull request by HarrisonLeach1**
GitHub Issue (If applicable): #27 

## PR Type

What kind of change does this PR introduce?
- Feature

## What is the current behavior?
- No interaction with the Reddit API

## What is the new behavior?
 - Reddit API data is consumed and can be accessed on the backend

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tested code with current upstream repo
- [x] Added new tests to the test suite for the feature
- [x] Tested the full test suite, and ensured that the feature does not break current build.
- [ ] Docs have been added/updated which fit (for bug fixes / features)
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal)

## Other information
- An `application.properties` file has been created, all credentials and other configuration properties should be added here. 
- Mockito has been added to mock out dependencies for unit testing. For more information about Mockito see [here](https://site.mockito.org/).
- JRAW has been used for making api requests, unfortunately it is a blocking request. However, the api client can be changed in future if needed as there is little code required. 

### For local testing
- The application won't work immediately after you pull the changes because the `application.properties` has empty configuration values. 
- What you should do is:
1. Make a copy of `application.properties` and rename it to `application-dev.properties` in the same directory.

It should look like: 
```
providers.reddit.username=
providers.reddit.password=
providers.reddit.clientId=
providers.reddit.clientSecret=
```
2. Fill in the properties with values, a usable example has been provided in the secure communications channel.
3. Run the program with the command-line arguments: `-Dspring.profiles.active=dev`. If using IntelliJ Go to ServerApplication > Edit Configurations and add `-Dspring.profiles.active=dev` to VM Options.

### CODE REVIEW

File: server/.gitignore [hunk 1]
```diff
@@ -29,3 +29,6 @@ build/
 
 ### VS Code ###
 .vscode/
+
+### Development variables/secrets ###
+application-dev.properties
```
Comment (jordansimsmith) [inline server/.gitignore]: Nice

File: server/src/main/java/com/feeder/server/controllers/HelloController.java [hunk 2]
```diff
@@ -1,16 +1,30 @@
...
 
+import java.util.List;
+
 @RestController
 @RequestMapping("/hello")
 public class HelloController {
 
+    private static final Logger logger = LoggerFactory.getLogger(HelloController.class);
+    @Autowired private ApplicationProperties properties;
+    @Autowired private RedditFeedProvider redditFeedProvider;
```
Comment (jordansimsmith) [inline server/src/main/java/com/feeder/server/controllers/HelloController.java]: Nice use of DI here

**Comment by HarrisonLeach1** [comment]
A lot of the implementation around provider implementation i.e. the data type, method signature, only 1 page is retrieved, are all due to change. I thought it would be good to get this PR up sooner than later to give an idea of how API secrets can be stored.

File: server/src/test/java/com/feeder/server/providers/reddit/RedditFeedProviderTest.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import org.slf4j.Logger;
+
+import java.util.Arrays;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.ArgumentMatchers.anyInt;
+import static org.mockito.Mockito.*;
+
+class RedditFeedProviderTest {
```
Comment (jordansimsmith) [inline server/src/test/java/com/feeder/server/providers/reddit/RedditFeedProviderTest.java]: personally, I would have public modifiers on the class as well as the test cases (more common for what I've seen).

Not blocking, however, and Im willing to hear other opinions on this.

File: server/src/test/java/com/feeder/server/providers/reddit/RedditFeedProviderTest.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    private TestableRedditFeedProvider subject;
+    private RedditClient mockRedditClient;
```
Comment (jordansimsmith) [inline server/src/test/java/com/feeder/server/providers/reddit/RedditFeedProviderTest.java]: You can use the mockito test runner, to use the `@Mock` annotation to automatically wire up the mocks.

File: server/src/test/java/com/feeder/server/providers/reddit/RedditFeedProviderTest.java (hunk 3 above)
Comment (HarrisonLeach1) [reply server/src/test/java/com/feeder/server/providers/reddit/RedditFeedProviderTest.java]: Oops, I didn't see this

File: server/src/main/resources/application.properties [hunk 5]
```diff
@@ -1 +1,4 @@
-
+providers.reddit.username=
+providers.reddit.password=
+providers.reddit.clientId=
+providers.reddit.clientSecret=
```
Comment (Minus20Five) [inline server/src/main/resources/application.properties]: Are we planning to support interactive oauth flow?

Comment (HarrisonLeach1) [reply server/src/main/resources/application.properties]: It was discussed in the meeting that we would not need to for the initial MVP. But a stretch goal is that we may be able to get it working for some providers.

File: server/pom.xml [hunk 6]
```diff
@@ -48,6 +48,12 @@
 			<version>3.3.0</version>
 			<scope>test</scope>
 		</dependency>
+
+		<dependency>
+			<groupId>junit</groupId>
+			<artifactId>junit</artifactId>
+			<scope>test</scope>
+		</dependency>
```
Comment (jordansimsmith) [inline server/pom.xml]: As discussed, lets replace the JUnit 4 `@RunWith` annotatation with the JUnit 5 counterpart, thus removing the need for both junit 4 and 5.

File: server/src/main/resources/application.properties (hunk 5 above)
Comment (Minus20Five) [reply server/src/main/resources/application.properties]: Yup, just double checking.

File: server/src/test/java/com/feeder/server/providers/reddit/RedditFeedProviderTest.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    private class TestableRedditFeedProvider extends RedditFeedProvider {
+        public TestableRedditFeedProvider(ApplicationProperties mockProperties) {
+            super(mockProperties);
+        }
+
+        @Override
+        protected RedditClient getAuthenticatedClient(NetworkAdapter adapter, Credentials credentials) {
+            return mockRedditClient;
+        }
+    }
```
Comment (Minus20Five) [inline server/src/test/java/com/feeder/server/providers/reddit/RedditFeedProviderTest.java]: Super-nit.
You can refactor complex constructor code in `RedditFeedProvider` by putting the creation logic in a tiny factory.

**Comment by AlexanderTheGrape** [comment]
Seeing the commit history, it appears the commits have not been squashed. The assignment requires us to do this before a merge, unfortunately: 
"Prior to merging, all commits should be squashed. The reviewer should request the
code author to do this if required. "

File: server/src/main/java/com/feeder/server/controllers/HelloController.java [hunk 8]
```diff
@@ -1,16 +1,30 @@
...
+
     @GetMapping
     public Hello readHello() {
+
+        List items = redditFeedProvider.getFeed(2);
```
[the remark this answers is not in the document]
Comment (HarrisonLeach1) [reply server/src/main/java/com/feeder/server/controllers/HelloController.java]: I originally had this but IntelliJ told me to change it, i'll change it back.

File: server/src/main/java/com/feeder/server/providers/reddit/RedditFeedProvider.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+import net.dean.jraw.oauth.OAuthHelper;
+import net.dean.jraw.pagination.Paginator;
+import org.springframework.stereotype.Service;
+import java.util.List;
+
+/***
+ * A RedditFeedProvider is responsible for retrieving Reddit posts from a users homepage from the Reddit API.
+ */
+@Service
+public class RedditFeedProvider {
```
[the remark this answers is not in the document]
Comment (HarrisonLeach1) [reply server/src/main/java/com/feeder/server/providers/reddit/RedditFeedProvider.java]: I was thinking of doing this once the backend integration team had defined their interfaces and data types.

File: server/src/test/java/com/feeder/server/providers/reddit/RedditFeedProviderTest.java (hunk 7 above)
Comment (HarrisonLeach1) [reply server/src/test/java/com/feeder/server/providers/reddit/RedditFeedProviderTest.java]: I was also suggesting to make a ProviderFactory once the integration team had made their interfaces.

File: server/src/main/java/com/feeder/server/providers/reddit/RedditFeedProvider.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (HarrisonLeach1) [reply server/src/main/java/com/feeder/server/providers/reddit/RedditFeedProvider.java]: Im thinking I might as well add it now, what is our interface naming convention? is IProvider good?

**Comment by HarrisonLeach1** [comment]
> `mvn test` is failing for me because of some configuration issue with `ApplicationProperties`

My bad I will fix it so it doesn't require application.properties in the case of unit tests.

It'll be because of empty application properties in the build artifact, as explained in the **For local testing** section. And you'll need to run it with the dev profile. e.g. `mvn test -Dspring.profiles.active=dev`

File: server/src/main/java/com/feeder/server/providers/reddit/RedditFeedProvider.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (HarrisonLeach1) [reply server/src/main/java/com/feeder/server/providers/reddit/RedditFeedProvider.java]: What do you think our naming convention for interfaces should be? for example IProvider?

**Comment by HarrisonLeach1** [comment]
> `mvn test` is failing for me because of some configuration issue with `ApplicationProperties`

I have fixed the issues, let me know if is working for you.

File: server/src/main/java/com/feeder/server/providers/reddit/RedditFeedProvider.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (Minus20Five) [reply server/src/main/java/com/feeder/server/providers/reddit/RedditFeedProvider.java]: IProvider seems to be c# convention. Provider<DataType> is more Java-like imo.

**Pull request merged**